agent/proxycfg/upstreams.go

```diff
@@ -92,55 +90,60 @@ func (s *handlerUpstreams) handleUpdateUpstreams(ctx context.Context, u cache.Up
 		}
 		upstreamsSnapshot.WatchedUpstreamEndpoints[uid][targetID] = resp.Nodes
 
-		var passthroughAddrs map[string]ServicePassthroughAddrs
+		if s.kind != structs.ServiceKindConnectProxy || s.proxyCfg.Mode != structs.ProxyModeTransparent {
+			return nil
+		}
 
-		for _, node := range resp.Nodes {
-			if snap.Proxy.Mode == structs.ProxyModeTransparent && node.Service.Proxy.TransparentProxy.DialedDirectly {
-				if passthroughAddrs == nil {
-					passthroughAddrs = make(map[string]ServicePassthroughAddrs)
+		if _, ok := upstreamsSnapshot.PassthroughUpstreams[uid]; ok {
+			for addr := range upstreamsSnapshot.PassthroughUpstreams[uid][targetID] {
+				if indexed := upstreamsSnapshot.PassthroughIndices[addr]; indexed.targetID == targetID && indexed.upstreamID.Matches(uid) {
+					delete(upstreamsSnapshot.PassthroughIndices, addr)
 				}
+			}
+			upstreamsSnapshot.PassthroughUpstreams[uid][targetID] = make(map[string]struct{})
+		}
 
-				svc := node.Service.CompoundServiceName()
-
-				// Overwrite the name if it's a connect proxy (as opposed to Connect native).
-				// We don't reference the proxy name directly for things like SNI, but rather the name
-				// of the destination. The enterprise meta of a proxy will always be the same as that of
-				// the destination service, so that remains intact.
-				if node.Service.Kind == structs.ServiceKindConnectProxy {
-					dst := node.Service.Proxy.DestinationServiceName
-					if dst == "" {
-						dst = node.Service.Proxy.DestinationServiceID
-					}
-					svc.Name = dst
-				}
+		passthroughs := make(map[string]struct{})
+
+		for _, node := range resp.Nodes {
+			if !node.Service.Proxy.TransparentProxy.DialedDirectly {
+				continue
+			}
 
-				sni := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), svc.PartitionOrDefault(), snap.Datacenter, snap.Roots.TrustDomain)
+			// Make sure to use an external address when crossing partitions.
+			// Datacenter is not considered because transparent proxies cannot dial other datacenters.
```

Comment:
Should you somehow explicitly skip this if there's a datacenter mismatch instead of just letting it do something weird?

Reply:
Hmm looking at this again it is possible for a tproxy to have an upstream target in another DC. A transparent proxy's upstreams are inferred from intentions, but one of those upstreams from intentions could configure a redirect to a different DC. 

So I think the right thing here is to not skip, but rather to incorporate a DC check. Made that change.